File: code/java/399.java
class Solution {
    private class UnionFind{
        private int[] parent;
        private double[] weight;
        public UnionFind(int n){
            this.parent = new int[n];
            this.weight = new double[n];
            for(int i=0;i<n;i++){
                parent[i] = i;
                weight[i] = 1.0d; // 边的权重全部初始化为1(自己指向自己，x/x=1)
            }
        }
        public int find(int x){
            if(x!=parent[x]){
                // 查询以及路径压缩
                int origin = parent[x];
                // 递归查找，一定会查找到最顶上的节点，所以就将该路径下的所有节点拼接到最顶上那个节点下方
                parent[x] = find(parent[x]);
                weight[x]*=weight[origin];
            }
            return parent[x];
        }
        public void union(int p,int q,double value){
            int rootP = find(p);
            int rootQ = find(q);
            if(rootP==rootQ) return;
            parent[rootP] = rootQ;
            weight[rootP] = weight[q]*value/weight[p];
        }
        public double isConnected(int p,int q){
            int rootP = find(p);
            int rootQ = find(q);
            if(rootP==rootQ){
                return weight[p]/weight[q];
            }else{
                return -1.0d;
            }
        }
    }

    public double[] calcEquation(List<List<String>> equations, double[] values, List<List<String>> queries) {
        int size = equations.size();
        UnionFind uf = new UnionFind(2*size);
        Map<String,Integer> hashMap = new HashMap<>(2*size);
        int id = 0; // 记录有效符号个数
        // step1. 建立有向图
        for(int i=0;i<size;i++){
            List<String> equation = equations.get(i);
            String var1 = equation.get(0);
            String var2 = equation.get(1);
            if(!hashMap.containsKey(var1)){
                hashMap.put(var1,id);
                id++;
            }
            if(!hashMap.containsKey(var2)){
                hashMap.put(var2,id);
                id++;
            }
            uf.union(hashMap.get(var1),hashMap.get(var2),values[i]);
        }
        // step2. 求解
        int queriesSize = queries.size();
        double[] res = new double[queriesSize];
        for (int i = 0; i < queriesSize; i++) {
            String var1 = queries.get(i).get(0);
            String var2 = queries.get(i).get(1);

            Integer id1 = hashMap.get(var1);
            Integer id2 = hashMap.get(var2);

            if (id1 == null || id2 == null) {
                res[i] = -1.0d;
            } else {
                res[i] = uf.isConnected(id1, id2);
            }
        }
        return res;
    }
}

File: code/python/399.py
class UF:
    def __init__(self,n):
        self.parent = [i for i in range(n)]
        self.values = [1]*n
    def find(self,x):
        if x!=self.parent[x]:
            origin = self.parent[x]
            self.parent[x] = self.find(self.parent[x])
            self.values[x]*=self.values[origin]
        return self.parent[x]
    def union(self,x,y,value):
        rootX = self.find(x)
        rootY = self.find(y)
        if rootX==rootY:
            return
        self.parent[rootX] = rootY
        self.values[rootX] = self.values[y]*value/self.values[x]
    def isConnected(self,x,y):
        rootX = self.find(x)
        rootY = self.find(y)
        if rootX==rootY:
            return self.values[x]/self.values[y]
        return -1
class Solution:
    def calcEquation(self, equations: List[List[str]], values: List[float], queries: List[List[str]]) -> List[float]:
        charIndex = {}
        uf = UF(len(equations)*2)
        index = 0
        for i in range(len(equations)):
            var1 = equations[i][0]
            var2 = equations[i][1]
            if var1 not in charIndex:
                charIndex[var1] = index
                index+=1
            if var2 not in charIndex:
                charIndex[var2] = index
                index+=1
            uf.union(charIndex[var1],charIndex[var2],values[i])
        res = []
        for i in range(len(queries)):
            var1 = queries[i][0]
            var2 = queries[i][1]

            if var1 not in charIndex or var2 not in charIndex:
                res.append(-1)
                continue
            res.append(uf.isConnected(charIndex[var1],charIndex[var2]))
        return res